usersession/userd/privileged_desktop_launcher.go

```diff
@@ -0,0 +1,298 @@
+}
+
+// desktopFileIDToFilename determines the path associated with a desktop file ID.
+func desktopFileIDToFilename(desktopFileExists fileExists, desktopFileID string) (string, error) {
+
+	// OpenDesktopEntry() currently only supports launching snap applications from
+	// desktop files in /var/lib/snapd/desktop/applications and these desktop files are
+	// written by snapd and considered safe for userd to process.
+	// Since we only access /var/lib/snapd/desktop/applications, ignore
+	// https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
+	baseDir := dirs.SnapDesktopFilesDir
+
+	desktopFile, err := findDesktopFile(desktopFileExists, baseDir, strings.Split(desktopFileID, "-"))
+
+	if err == nil {
+		return desktopFile, nil
+	}
+
+	return "", fmt.Errorf("cannot find desktop file for %q", desktopFileID)
+}
+
+// verifyDesktopFileLocation checks the desktop file location:
+// we only consider desktop files in dirs.SnapDesktopFilesDir
+func verifyDesktopFileLocation(desktopFile string) error {
+	if !strings.HasPrefix(desktopFile, dirs.SnapDesktopFilesDir+"/") {
+		// We currently only support launching snap applications from desktop files in
+		// /var/lib/snapd/desktop/applications and these desktop files are written by snapd and
+		// considered safe for userd to process. If other directories are added in the future,
+		// verifyDesktopFileLocation() and parseExecCommand() may need to be updated.
+		return fmt.Errorf("only launching snap applications from /var/lib/snapd/desktop/applications is supported")
+	}
+
+	if filepath.Clean(desktopFile) != desktopFile {
+		return fmt.Errorf("desktop file has unclean path: %q", desktopFile)
+	}
+
+	return nil
+}
+
+// readExecCommandFromDesktopFile parses the desktop file to get the Exec entry and
+// checks that the BAMF_DESKTOP_FILE_HINT is present and refers to the desktop file.
+func readExecCommandFromDesktopFile(desktopFile string) (exec string, icon string, err error) {
+	file, err := os.Open(desktopFile)
+	if err != nil {
+		return exec, icon, err
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+
+	in_desktop_section := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		if line == "[Desktop Entry]" {
+			in_desktop_section = true
+		} else if strings.HasPrefix(line, "[Desktop Action ") {
+			// maybe later we'll add support here
+			in_desktop_section = false
+		} else if strings.HasPrefix(line, "[") {
+			in_desktop_section = false
+		} else if in_desktop_section {
+			if strings.HasPrefix(line, "Exec=") {
+				exec = strings.TrimPrefix(line, "Exec=")
+			} else if strings.HasPrefix(line, "Icon=") {
+				icon = strings.TrimPrefix(line, "Icon=")
+			}
+		}
+	}
+
+	expectedPrefix := fmt.Sprintf("env BAMF_DESKTOP_FILE_HINT=%s "+dirs.SnapBinariesDir, desktopFile)
+	if !strings.HasPrefix(exec, expectedPrefix) {
+		return "", "", fmt.Errorf("Desktop file %q has an unsupported 'Exec' value: %q", desktopFile, exec)
```

Comment:
errors by convention are mostly lower case to deal with errors wrapping errors:  s/Desktop/desktop/ here